src/python/zquantum/core/graph.py

```diff
@@ -1,43 +1,42 @@
+import itertools
 import json
 import random
-from random import uniform
-from typing import Dict, Optional, cast
+import warnings
+from typing import Iterator, Optional
 
 import networkx as nx
-from zquantum.core.typing import AnyPath, LoadSource
 
+from .serialization import ensure_open
+from .typing import DumpTarget, LoadSource
 from .utils import SCHEMA_VERSION
 
+GRAPH_SCHEMA = SCHEMA_VERSION + "-graph"
 
-def save_graph(graph: nx.Graph, filename: AnyPath):
+
+def save_graph(graph: nx.Graph, filename: DumpTarget):
     """Saves a NetworkX graph object to JSON file.
 
     Args:
-        graph (networks.Graph): the input graph object
-        filename (string): name of the output file
+        graph: the input graph object
+        filename: name of the output file
     """
-    f = open(filename, "w")
-    graph_dict = nx.readwrite.json_graph.node_link_data(graph)
-    graph_dict["schema"] = SCHEMA_VERSION + "-graph"
-    json.dump(graph_dict, f, indent=2)
-    f.close()
+    with ensure_open(filename, "w") as f:
+        graph_dict = {
+            **nx.readwrite.json_graph.node_link_data(graph),
+            "schema": GRAPH_SCHEMA,
+        }
+        json.dump(graph_dict, f, indent=2)
```

Comment:
Shall we keep the `f.close()` line?

Reply:
Not really – the `open` context manager (wrapped by `ensure_open`) closes the file on exiting the block 🙂